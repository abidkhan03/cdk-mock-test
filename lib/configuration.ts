export const CARMATECH_CONFIG = {
    Prod: {
        ARN: 'arn:aws:codestar-connections:us-east-2:395929101814:connection/7d77688a-1802-456a-9a06-81f53412e887',
        ACCOUNT_ID: '395929101814',
        REGION: 'us-east-2',
    }
}

export const ALPACA_CONFIG = {
    Prod: {
        ALPACA_API_KEY: 'AK4ECDTG3S4N42CPFDG0',
        ALPACA_API_SECRET: 'fzOyObTExs70mDWYaeh2fE6oUA0CmWyO7dZukAIn',
    }
}


// import { CfnCapabilities, Duration, RemovalPolicy, SecretValue, Stack, StackProps, Stage } from "aws-cdk-lib";
// import { Artifact, Pipeline } from "aws-cdk-lib/aws-codepipeline";
// import { CloudFormationCreateReplaceChangeSetAction, CloudFormationCreateUpdateStackAction, CloudFormationExecuteChangeSetAction, CodeBuildAction, CodeStarConnectionsSourceAction, GitHubSourceAction, GitHubTrigger } from "aws-cdk-lib/aws-codepipeline-actions";
// import { Construct } from "constructs";
// import { CARMATECH_CONFIG } from "./configuration";
// import { BuildEnvironmentVariableType, BuildSpec, LinuxBuildImage, PipelineProject } from "aws-cdk-lib/aws-codebuild";
// import { Bucket } from "aws-cdk-lib/aws-s3";
// import { Effect, ManagedPolicy, Policy, PolicyStatement } from "aws-cdk-lib/aws-iam";
// import { Function, Code, Runtime } from "aws-cdk-lib/aws-lambda";


// export class DailyFoodNotificationLambdaPipeline extends Stack {
//   constructor(scope: Construct, id: string, props: StackProps) {
//       super(scope, id, props);

//       // Creates an S3 bucket
//     const bucket = new Bucket(this, 'MyBucket', {
//         bucketName: 'daily-menu-items-bucket',
//         removalPolicy: RemovalPolicy.DESTROY
//       });

//       const myLambda = new Function(this, 'MyLambdaHandler', {
//         code: Code.fromAsset('./misc_cdk/', {
//             bundling: {
//                 image: Runtime.PYTHON_3_9.bundlingImage,
//                 command: [
//                     'bash', '-c',
//                     'pip install -r requirements.txt -t /asset-output && cp -au . /asset-output'
//                 ],
//             },
//         }),
//         runtime: Runtime.PYTHON_3_9,
//         handler:'lambda_function.lambda_handler', // Exported handler name
//         timeout: Duration.seconds(300),
//         environment: {
//             BUCKET_NAME: bucket.bucketName,
//         }
//       });

//     //   bucket.grantReadWrite(myLambda);
        

//       //Creating s3 Bucket
//       const artifactsBucket = new Bucket(this, "S3BucketForPipelineArtifacts");

//       //Codepipeline
//       const codepipeline = new Pipeline(this, 'CodePipelineForLambdaDeployment', {});

//       //Source Stage
//       const sourceArtifact = new Artifact();

//       const sourceAction = new GitHubSourceAction({
//         actionName: 'Lambda_Source',
//         owner: 'abdullah5abid',
//         repo: 'misc_cdk',
//         output: sourceArtifact,
//         branch: 'master', //change to your branch name
//         oauthToken: SecretValue.secretsManager('github-token'),
//         trigger: GitHubTrigger.WEBHOOK,
//       });

//       codepipeline.addStage({
//           stageName: "Source",
//           actions: [sourceAction],
//       });

//       //Build Stage
//       const buildArtifact = new Artifact();
//       const buildAction = new CodeBuildAction({
//           actionName: "BuildAction",
//           input: sourceArtifact,
//           project: this.createCodeBuildProject(artifactsBucket.bucketName),
//           outputs: [buildArtifact]
//       });

//       codepipeline.addStage({
//           stageName: "Build",
//           actions: [buildAction],
//       }
//       );

//       //Deploy Stage
//       const stackName = 'Codepipeline-Lambda-Stack';
//       const changeSetName = 'StagedChangeSet'

//       const createReplaceChangeSetAction = new CloudFormationCreateReplaceChangeSetAction({
//           actionName: "PrepareChanges",
//           stackName: stackName,
//           changeSetName: changeSetName,
//           templatePath: buildArtifact.atPath('outputtemplate.yml'),
//           cfnCapabilities: [
//               CfnCapabilities.NAMED_IAM,
//               CfnCapabilities.AUTO_EXPAND
//           ],
//           adminPermissions: false,
//           runOrder: 1
//       });

//       const executeChangeSetAction = new CloudFormationExecuteChangeSetAction({
//           actionName: "ExecuteChanges",
//           changeSetName: changeSetName,
//           stackName: stackName,
//           runOrder: 2
//       })

//       codepipeline.addStage({
//           stageName: "Deploy",
//           actions: [
//               createReplaceChangeSetAction,
//               executeChangeSetAction
//           ],
//       }
//       );

//       //Permission for CloudFormation to access Lambda and other resources
//       createReplaceChangeSetAction.deploymentRole.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('AWSLambdaExecute'));
//       createReplaceChangeSetAction.deploymentRole.attachInlinePolicy(this.getCodePipelineCloudFormationInlinePolicy());
//   }

//   //Creating code build project
//   private createCodeBuildProject = (artifactsBucket: string): PipelineProject => {
//       const codeBuildProject = new PipelineProject(this, 'CodeBuildProject', {
//           projectName: 'Notification-Lambda',
//           environment: {
//               buildImage: LinuxBuildImage.STANDARD_5_0
//           },
//           buildSpec: BuildSpec.fromObject(this.getBuildSpecContent(artifactsBucket))
//       });

//       codeBuildProject.role?.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('AmazonS3FullAccess'));
//       return codeBuildProject;
//   }

//   //Creating the build spec content.
//   private getBuildSpecContent = (artifactsBucket: string) => {
//       return {
//           version: '0.2',
//           phases: {
//               install: {
//                   'runtime-versions': { python: '3.9' },
//                   commands: [
//                       'pip install --target=./MISC -r MISC/requirements.txt'
//                   ]
//               },
//               pre_build: {
//                   commands: [
//                       'echo Pre-build completed'
//                   ]
//               },
//               build: {
//                   commands: [
//                       'echo Build started on `date`',
//                       'export BUCKET=' + artifactsBucket,
//                       'sam package --template-file buildspec.yml --s3-bucket $BUCKET --output-template-file outputtemplate.yml',
//                       'echo Build completed on `date`'
//                   ]
//               }
//           },
//           artifacts: {
//               type: 'zip',
//               files: [
//                   'buildspec.yml',
//                   'outputtemplate.yml'
//               ]
//           }
//       }
//   };

//   //Inline permission policy for CloudFormation
//   private getCodePipelineCloudFormationInlinePolicy = () => {
//       return new Policy(this, 'CodePipelineCloudFormationInlinePolicy', {
//           statements: [
//               new PolicyStatement({
//                   effect: Effect.ALLOW,
//                   actions: [
//                       "apigateway:*",
//                       "codedeploy:*",
//                       "lambda:*",
//                       "cloudformation:CreateChangeSet",
//                       "iam:GetRole",
//                       "iam:CreateRole",
//                       "iam:DeleteRole",
//                       "iam:PutRolePolicy",
//                       "iam:AttachRolePolicy",
//                       "iam:DeleteRolePolicy",
//                       "iam:DetachRolePolicy",
//                       "iam:PassRole",
//                       "s3:*",
//                     //   "s3:GetObjectVersion",
//                     //   "s3:GetBucketVersioning",
//                     //   "s3:PutObject",
//                     //   "s3:PutObjectVersion",
//                     //   "s3:DeleteObject",
//                     //   "s3:DeleteObjectVersion",
//                     //   "s3:ListBuckets",
//                   ],
//                   resources: ['*']
//               })
//           ]
//       })
//   }

// }