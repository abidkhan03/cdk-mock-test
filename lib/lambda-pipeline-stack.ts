import { Construct } from 'constructs';
import { CfnCapabilities, CfnOutput, Duration, SecretValue, Stack, StackProps } from 'aws-cdk-lib';
import { Artifact, Pipeline } from 'aws-cdk-lib/aws-codepipeline';
import { 
  CodeBuildAction, 
  CloudFormationCreateReplaceChangeSetAction, 
  CloudFormationExecuteChangeSetAction, 
  GitHubSourceAction, 
  GitHubTrigger, 
  CodeStarConnectionsSourceAction
} from 'aws-cdk-lib/aws-codepipeline-actions';
import { ManagedPolicy, Policy, PolicyStatement, Effect, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { BuildSpec, LinuxBuildImage, PipelineProject } from 'aws-cdk-lib/aws-codebuild';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { CARMATECH_CONFIG } from './configuration';
import { Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';
import { LambdaIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';


export class DailyFoodNotificationLambdaPipeline extends Stack {
  constructor(scope: Construct, id: string, props: StackProps) {
      super(scope, id, props);

      //Creating s3 Bucket
      const artifactsBucket = new Bucket(this, "S3BucketForPipelineArtifacts");

    //   //Create IAM Role for Lambda Function
    //   const lambdaRole = new Role(this, "lambdaRole", {
    //     assumedBy: new ServicePrincipal('lambda.amazonaws.com'),
    //     managedPolicies: [
    //         ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole'),
    //         ManagedPolicy.fromAwsManagedPolicyName('AmazonS3FullAccess')
    //     ]
    // });

      //Codepipeline
      const codepipeline = new Pipeline(this, 'CodePipelineForLambdaDeployment', {});

      //Source Stage
      const sourceArtifact = new Artifact();

      const sourceAction = new GitHubSourceAction({
        actionName: 'Lambda_Source',
            owner: 'abdullah5abid',
            repo: 'misc_cdk',
            output: sourceArtifact,
            branch: 'master',
            oauthToken: SecretValue.secretsManager('github-token'),
            trigger: GitHubTrigger.WEBHOOK,
      });

      codepipeline.addStage({
          stageName: "Source",
          actions: [sourceAction],
      });

      //Build Stage
      const buildArtifact = new Artifact();
      const buildAction = new CodeBuildAction({
          actionName: "BuildAction",
          input: sourceArtifact,
          project: this.createCodeBuildProject(artifactsBucket.bucketName),
          outputs: [buildArtifact]
      });

      codepipeline.addStage({
          stageName: "Build",
          actions: [buildAction],
      }
      );

      //Deploy Stage
      const stackName = 'Codepipeline-Lambda-Stack';
      const changeSetName = 'StagedChangeSet'

      const createReplaceChangeSetAction = new CloudFormationCreateReplaceChangeSetAction({
          actionName: "PrepareChanges",
          stackName: stackName,
          changeSetName: changeSetName,
          templatePath: buildArtifact.atPath('outputtemplate.yml'),
          cfnCapabilities: [
              CfnCapabilities.NAMED_IAM,
              CfnCapabilities.AUTO_EXPAND
          ],
          adminPermissions: false,
          runOrder: 1
      });

      const executeChangeSetAction = new CloudFormationExecuteChangeSetAction({
          actionName: "ExecuteChanges",
          changeSetName: changeSetName,
          stackName: stackName,
          runOrder: 2
      })

      codepipeline.addStage({
          stageName: "Deploy",
          actions: [
              createReplaceChangeSetAction,
              executeChangeSetAction
          ],
      }
      );

    //   //Create lambda function
    //   const dailyMenuLambda = new Function(this, "dailyLunchMenu", {
    //     functionName: "DailyMenuNotificationLambda",
    //     runtime: Runtime.PYTHON_3_9,
    //     code: Code.fromBucket(artifactsBucket, "MISC/lambda_function.zip"),
    //     handler: 'lambda_function.lambda_handler',
    //     timeout: Duration.seconds(200),
    //     role: lambdaRole,
    //   });

    //   //Create API Gateway
    //   const api = new RestApi(this, "ApiGateway");

    //   //Create API Gateway resource and method
    //   const lambdaIntegration = new LambdaIntegration(dailyMenuLambda);
    //   api.root.addResource('dailylunchmenu').addMethod('GET', lambdaIntegration);

    //   // Output the API Gateway endpoint URL
    //   new CfnOutput(this, 'ApiGatewayEndpoint', {
    //     value: api.url ?? 'Something went wrong',
    //   })

      //Permission for CloudFormation to access Lambda and other resources
      createReplaceChangeSetAction.deploymentRole.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('AWSLambdaExecute'));
      createReplaceChangeSetAction.deploymentRole.attachInlinePolicy(this.getCodePipelineCloudFormationInlinePolicy());
  }

  //Creating code build project
  private createCodeBuildProject = (artifactsBucket: string): PipelineProject => {
      const codeBuildProject = new PipelineProject(this, 'CodeBuildProject', {
          projectName: 'Notification-Lambda',
          environment: {
              buildImage: LinuxBuildImage.STANDARD_5_0
          },
          buildSpec: BuildSpec.fromObject(this.getBuildSpecContent(artifactsBucket))
      });

      codeBuildProject.role?.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('AmazonS3FullAccess'));
      return codeBuildProject;
  }

  //Creating the build spec content.
  private getBuildSpecContent = (artifactsBucket: string) => {
      return {
          version: '0.2',
          phases: {
              install: {
                  'runtime-versions': { python: '3.9' },
                  commands: [
                      'pip install --target=./MISC -r MISC/requirements.txt'
                  ]
              },
              pre_build: {
                  commands: [
                      'echo Pre-build completed'
                  ]
              },
              build: {
                  commands: [
                      'echo Build started on `date`',
                      'export BUCKET=' + artifactsBucket,
                      'sam package --template-file buildspec.yml --s3-bucket $BUCKET --output-template-file outputtemplate.yml',
                      'echo Build completed on `date`'
                  ]
              }
          },
          artifacts: {
              type: 'zip',
              files: [
                  'buildspec.yml',
                  'outputtemplate.yml',
                  'lambda_function.zip'
              ]
          }
      }
  };

  //Inline permission policy for CloudFormation
  private getCodePipelineCloudFormationInlinePolicy = () => {
      return new Policy(this, 'CodePipelineCloudFormationInlinePolicy', {
          statements: [
              new PolicyStatement({
                  effect: Effect.ALLOW,
                  actions: [
                      "apigateway:*",
                      "codedeploy:*",
                      "lambda:*",
                      "cloudformation:CreateChangeSet",
                      "iam:GetRole",
                      "iam:CreateRole",
                      "iam:DeleteRole",
                      "iam:PutRolePolicy",
                      "iam:AttachRolePolicy",
                      "iam:DeleteRolePolicy",
                      "iam:DetachRolePolicy",
                      "iam:PassRole",
                      "s3:GetObject",
                      "s3:GetObjectVersion",
                      "s3:GetBucketVersioning"
                  ],
                  resources: ['*']
              })
          ]
      })
  }

}
