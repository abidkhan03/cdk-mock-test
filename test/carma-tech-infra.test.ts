import { StockUltimateStack } from "../lib/stock_ultimate-stack";
import { expect as expectCDK, haveResource } from "@aws-cdk/assert";
import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import 'dotenv/config';
import { CARMATECH_CONFIG } from "../lib/configuration";

jest.mock('../lib/configuration');

jest.mock('../lib/stock_ultimate-stack', () => {
    return {
        StockUltimateStack: jest.fn()
    }
});

const app = new App();
const stockStack = new StockUltimateStack(app, 'StockTestStack', {
    env: {
        account: CARMATECH_CONFIG.Prod.ACCOUNT_ID,
        region: CARMATECH_CONFIG.Prod.REGION,
    },
});

beforeAll(() => {
    jest.spyOn(process, 'env', 'get').mockReturnValue({
        ACCOUNT_ID: '395929101814',
        REGION: 'us-east-2',
    });

    jest.setSystemTime(new Date('2023-11-01T00:00:00.000Z'));
});


function expectStackToHaveResource(resourceType: string, properties?: any) {
    expectCDK(stockStack).to(haveResource(resourceType, properties));
}
describe('Stock Ultimate Stack', () => {
    const template = Template.fromStack(stockStack);
    it('snapshot test', () => {
        expect(template.toJSON()).toMatchSnapshot();
    });
    describe('Resources existence', () => {
        it('stack should have a bucket name', () => {
            expectStackToHaveResource('AWS::S3::Bucket', {
                BucketName: 'stock-ultimate-bucket'
            })
        });
        it('stack should have a lambda function', () => {
            expectStackToHaveResource('AWS::Lambda::Function', {
                FunctionName: 'StockUltimateFunction',
            })
        });
    })
});
