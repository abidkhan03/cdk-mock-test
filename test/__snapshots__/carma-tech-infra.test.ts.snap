// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stock Ultimate Stack snapshot test 1`] = `
{
  "Outputs": {
    "StateMachineArn": {
      "Value": {
        "Ref": "StateMachine2E01A3A5",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CustomJsonClassifier": {
      "Properties": {
        "JsonClassifier": {
          "JsonPath": "$[*]",
        },
      },
      "Type": "AWS::Glue::Classifier",
    },
    "GlueCrawlerRole4AA4B387": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "glue.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Role assumed by AWS Glue Crawler",
      },
      "Type": "AWS::IAM::Role",
    },
    "GlueCrawlerRoleDefaultPolicy44958AC9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "StocksData9AE50C02",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "StocksData9AE50C02",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "glue:CreateDatabase",
                "glue:UpdateDatabase",
                "glue:CreateTable",
                "glue:UpdateTable",
                "glue:CreatePartition",
                "glue:UpdatePartition",
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
                "s3:PutObject",
                "s3:GetObject",
                "s3:ListBucket",
                "glue:GetDatabase",
                "glue:GetTable",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "GlueCrawlerRoleDefaultPolicy44958AC9",
        "Roles": [
          {
            "Ref": "GlueCrawlerRole4AA4B387",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StateMachine2E01A3A5": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "StateMachineRoleDefaultPolicy433448B5",
        "StateMachineRole78EC082E",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"WaitCompile","States":{"WaitCompile":{"Type":"Wait","Seconds":1,"Next":"Invoke Step Function"},"Invoke Step Function":{"Next":"Job Complete?","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2},{"ErrorEquals":["Stock API Failed"],"IntervalSeconds":10,"MaxAttempts":3,"BackoffRate":2}],"Type":"Task","OutputPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "StockNamesLambda05A0EA7B",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"Job Complete?":{"Type":"Choice","Choices":[{"Variable":"$.statusCode","NumericEquals":400,"Next":"Job Failed"},{"Variable":"$.statusCode","NumericEquals":200,"Next":"InvokeS3PutObject"}],"Default":"Internal Error"},"Internal Error":{"Type":"Fail","Error":"Internal Error","Cause":"Internal Error"},"Job Failed":{"Type":"Fail","Error":"Stock API Failed","Cause":"Stock API Failed"},"InvokeS3PutObject":{"End":true,"Type":"Task","ResultPath":"$.Payload","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:s3:putObject","Parameters":{"Bucket":"",
              {
                "Ref": "StocksData9AE50C02",
              },
              "","Key":"stocks_data/stock_assets_1698785705647.json","Body.$":"$.body"}}},"TimeoutSeconds":300}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "StateMachineRole78EC082E",
            "Arn",
          ],
        },
        "StateMachineName": "StockUltimateStateMachine",
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "StateMachineEventsRoleDBCDECD1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineEventsRoleDefaultPolicyFB602CA9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "StateMachine2E01A3A5",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineEventsRoleDefaultPolicyFB602CA9",
        "Roles": [
          {
            "Ref": "StateMachineEventsRoleDBCDECD1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StateMachineRole78EC082E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "states.us-east-2.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicy433448B5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "StocksData9AE50C02",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "StockNamesLambda05A0EA7B",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "StockNamesLambda05A0EA7B",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "s3:PutObject",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineRoleDefaultPolicy433448B5",
        "Roles": [
          {
            "Ref": "StateMachineRole78EC082E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StockDatabase82C37E98": {
      "Properties": {
        "CatalogId": "395929101814",
        "DatabaseInput": {
          "Name": "stock-ultimate-dbb",
        },
      },
      "Type": "AWS::Glue::Database",
    },
    "StockNamesLambda05A0EA7B": {
      "DependsOn": [
        "StockNamesLambdaServiceRoleDefaultPolicyC8FEEBC6",
        "StockNamesLambdaServiceRoleD548ADF9",
      ],
      "Properties": {
        "Code": {
          "ImageUri": {
            "Fn::Sub": "395929101814.dkr.ecr.us-east-2.\${AWS::URLSuffix}/cdk-hnb659fds-container-assets-395929101814-us-east-2:b0d5c954dccf802d6e1fd138f94d5b5d19c570df643955b03aac832a9cbf488a",
          },
        },
        "Description": "Created Stock Ultimate Lambda on 2023-10-31T20:55:05.487Z",
        "Environment": {
          "Variables": {
            "ALPACA_API_KEY": "AK4ECDTG3S4N42CPFDG0",
            "ALPACA_API_SECRET": "fzOyObTExs70mDWYaeh2fE6oUA0CmWyO7dZukAIn",
            "BUCKET_NAME": {
              "Ref": "StocksData9AE50C02",
            },
          },
        },
        "FunctionName": "StockUltimateFunction",
        "MemorySize": 1024,
        "PackageType": "Image",
        "Role": {
          "Fn::GetAtt": [
            "StockNamesLambdaServiceRoleD548ADF9",
            "Arn",
          ],
        },
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "StockNamesLambdaServiceRoleD548ADF9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "StockNamesLambdaServiceRoleDefaultPolicyC8FEEBC6": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "StocksData9AE50C02",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "StocksData9AE50C02",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:PutObject",
                "s3:GetObject",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "StocksData9AE50C02",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StockNamesLambdaServiceRoleDefaultPolicyC8FEEBC6",
        "Roles": [
          {
            "Ref": "StockNamesLambdaServiceRoleD548ADF9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StockUltimateCrawler": {
      "Properties": {
        "Classifiers": [
          {
            "Ref": "CustomJsonClassifier",
          },
        ],
        "DatabaseName": {
          "Ref": "StockDatabase82C37E98",
        },
        "Name": "StockUltimateCrawler",
        "Role": {
          "Fn::GetAtt": [
            "GlueCrawlerRole4AA4B387",
            "Arn",
          ],
        },
        "Schedule": {
          "ScheduleExpression": "cron(14 14 ? * WED *)",
        },
        "SchemaChangePolicy": {
          "DeleteBehavior": "DEPRECATE_IN_DATABASE",
          "UpdateBehavior": "UPDATE_IN_DATABASE",
        },
        "TablePrefix": "stock-ultimate-",
        "Targets": {
          "S3Targets": [
            {
              "Path": {
                "Fn::Join": [
                  "",
                  [
                    "s3://",
                    {
                      "Ref": "StocksData9AE50C02",
                    },
                    "/assets/",
                  ],
                ],
              },
            },
          ],
        },
      },
      "Type": "AWS::Glue::Crawler",
    },
    "StocksData9AE50C02": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "BucketName": "stock-ultimate-bucket",
        "VersioningConfiguration": {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "StocksDataPolicy5FAFC471": {
      "Properties": {
        "Bucket": {
          "Ref": "StocksData9AE50C02",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "s3.amazonaws.com",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "StocksData9AE50C02",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "StocksData9AE50C02",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "glue.amazonaws.com",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "StocksData9AE50C02",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "StocksData9AE50C02",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Principal": {
                "Service": "athena.amazonaws.com",
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "StocksData9AE50C02",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "StocksData9AE50C02",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "WeeklyTriggerRuleForStateMachine3DAC32CD": {
      "Properties": {
        "ScheduleExpression": "cron(14 7 ? * WED *)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Ref": "StateMachine2E01A3A5",
            },
            "Id": "Target0",
            "RoleArn": {
              "Fn::GetAtt": [
                "StateMachineEventsRoleDBCDECD1",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
